File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        auto mid=midnode(head);
        auto head2=reverse(mid);
        while(head2->next){
            auto ne=head->next;
            auto ne2=head2->next;
            head->next=head2;
            head2->next=ne;
            head=ne;
            head2=ne2;
        }
    }
    ListNode *midnode(ListNode* head){
        ListNode *slow=head,*fast=head;
        while(head && head->next){
            slow=slow->next;
            head=head->next->next;
        }
        return slow;
    }
    ListNode *reverse(ListNode* head){
        ListNode *pre=NULL,*cur=head;
        while(cur){
            auto ne=cur->next;
            cur->next=pre;
            pre=cur;
            cur=ne;
        }
        return pre;
    }
};

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def midnode(self,head: Optional[ListNode]) -> None:
        slow=fast=head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        return slow
    def reverse(self,head: Optional[ListNode]) -> None:
        pre,cur=None,head
        while cur:
            ne=cur.next
            cur.next=pre
            pre=cur
            cur=ne
        return pre
            
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        mid=self.midnode(head)
        head2=self.reverse(mid)
        while head2.next:
            ne=head.next
            ne2=head2.next
            head.next=head2
            head2.next=ne
            head=ne
            head2=ne2
        